### PR TITLE
Anexia: multi-interface / multi-IP

### PR DESCRIPTION
**Pull request by LittleFox94**
**What this PR does / why we need it**:

Can now create multiple network interfaces, each with multiple addresses (one from each configured prefix), easily allowing Machines to have e.g. an IPv4 and IPv6 address.

**Which issue(s) this PR fixes**:

Fixes ANXKUBE-1151

**What type of PR is this?**
/kind feature

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
Anexia: extend network configuration, allowing multiple interfaces and multiple IPs per interface
```

**Documentation**:

```documentation
NONE
```

### CODE REVIEW

**Comment by LittleFox94** [comment]
PR only for internal review before giving this upstream; please don't merge this in our fork

Build available as `anx-cr.io/d41-mgrosch-dev/machine-controller:anexia-multi-ip-2`

**Comment by LittleFox94** [comment]
It's missing some things we probably want (I'll keep this list updated as I remember things):

* validate configured prefixes to all be in the configured VLAN

File: pkg/cloudprovider/provider/anexia/network_provisioning.go
Comment (nachtjasmin) [inline pkg/cloudprovider/provider/anexia/network_provisioning.go]: It's pretty hard to keep the context in the brain when reading those lines. Can we simplify the branches a bit, maybe by using a `switch` statement? I am thinking about sth along these lines:

```suggestion
	switch {
	// Case 1: we already have networks, but it differs from the config.
	case len(status.Networks) != 0 && len(status.Networks) != len(config.Networks):
		log.Warn("size of status.Networks != config.Networks, this should not happen in normal operation - ignoring existing status")
	// Case 2: There are networks configured and we have to setup them.
	case len(config.Networks) > 0:
		status.Networks = make([]anxtypes.NetworkStatus, len(config.Networks))
	// Case 3: Do nothing otherwise.
	default:
	}
```

File: pkg/cloudprovider/provider/anexia/network_provisioning.go
Comment (nachtjasmin) [inline pkg/cloudprovider/provider/anexia/network_provisioning.go:63]: Would be nice if we could leave a comment here why this `NICType` got chosen. :3

File: pkg/cloudprovider/provider/anexia/network_provisioning.go
Comment (nachtjasmin) [inline pkg/cloudprovider/provider/anexia/network_provisioning.go:82]: Same goes for this mutex. A little comment that it's used to prevent VM race conditions doesn't hurt IMO. (and yes, I know that it's a leftover from the previous code. let's fix it quickly when we touch it anyway. ^^)

File: pkg/cloudprovider/provider/anexia/types/types.go
Comment (nachtjasmin) [inline pkg/cloudprovider/provider/anexia/types/types.go:82]: Wouldn't this be a breaking API change? We do the migration from the old VLAN identifier, so I'd assume we leave the old state filled out as well with the information of the first VLAN.

File: pkg/cloudprovider/provider/anexia/network_provisioning.go
Comment (nachtjasmin) [inline pkg/cloudprovider/provider/anexia/network_provisioning.go]: A little nitpick, but `%d` fits better here.

```suggestion
					log.Warnf("size of status.Networks[%d].Addresses != config.Networks[%d].Prefixes, this should not happen in normal operation - ignoring existing status", netIndex, netIndex)
```

File: examples/anexia-machinedeployment.yaml
Comment (nachtjasmin) [inline examples/anexia-machinedeployment.yaml]: > needs to get one reserved IP address

That sentence implicates manual intervention from the operators. Wouldn't

> gets one IP address reserved

be a better description of the behavior?

File: examples/anexia-machinedeployment.yaml
Comment (nachtjasmin) [inline examples/anexia-machinedeployment.yaml]: Nitpicky, but sentences end with a dot and start with a big letter. :p

**Review by nachtjasmin — CHANGES_REQUESTED** [review]
I haven't done a fully-fledged end to end review (yet), this review is based on the code alone.

So far it looks pretty good, thanks a lot for the effort! I'd do the end-to-end review when the PR has left the draft state.

File: pkg/cloudprovider/provider/anexia/network_provisioning.go
Comment (nachtjasmin) [reply pkg/cloudprovider/provider/anexia/network_provisioning.go:82]: The ticket even got marked as completed, maybe we can remove the mutex completely?

File: pkg/cloudprovider/provider/anexia/types/types.go
Comment (AndromedaCLI) [reply pkg/cloudprovider/provider/anexia/types/types.go:82]: In ANXKUBE-1131 we also specifically want to keep the old way functional for migration, so I think keeping the old status is a good idea as well

File: pkg/cloudprovider/provider/anexia/network_provisioning.go
Comment (LittleFox94) [reply pkg/cloudprovider/provider/anexia/network_provisioning.go:63]: I agree, but it's not like I know this either - it's the one thing supported by ADC API :upside_down_face:

File: pkg/cloudprovider/provider/anexia/network_provisioning.go
Comment (LittleFox94) [reply pkg/cloudprovider/provider/anexia/network_provisioning.go:82]: The ticket got marked as completed, but if you read the last comment, it just acknowledges this still happening if two calls happen "nearly on the same millisecond" - I'm not happy with this, will add a comment though (but I just moved this code around ^^')

File: pkg/cloudprovider/provider/anexia/types/types.go
Comment (LittleFox94) [reply pkg/cloudprovider/provider/anexia/types/types.go:82]: This part of the status is not used anymore after the Machine is created, but only to allow one reconcile to reserve IPs that the following reconciles can then use to create the machine.

Keeping the status would make the code a lot more complex, for imho no real benefit (the only benefit is when updating machine-controller while a Machine is being created (IP already reserved but VM provisioning request not yet sent) - this new version cannot pick up from that and will reserve a new IP instead, the already reserved one will just timeout and be freed again by the Engine).

I did not test how this version of machine-controller behaves when reconciling an existing VM though - I would guess nothing happens, but I will check this.

> In ANXKUBE-1131 we also specifically want to keep the old way functional for migration, so I think keeping the old status is a good idea as well

The old way of configuring the MachineDeployment is still functional and translated internally in machine-controller to basically ...

```yaml
cloudProviderSpec:
  networks:
  - vlan: $cloudProviderSpec.vlanID
    prefixes:
    - "" # having an "empty-string prefix" tells the rest of the program to reserve IPs based on only the VLAN
```

File: pkg/cloudprovider/provider/anexia/network_provisioning.go
Comment (LittleFox94) [reply pkg/cloudprovider/provider/anexia/network_provisioning.go]: eh, that shouldn't be relevant, `%v` works just as well - using an explicit argument index instead of passing `netIndex` twice on the other hand .. ^^

File: pkg/cloudprovider/provider/anexia/network_provisioning.go
Comment (LittleFox94) [reply pkg/cloudprovider/provider/anexia/network_provisioning.go]: that's a different behavior though, irrelevant difference, but still

I also don't think this is more readable .. I will add a nice comment though, maybe that helps already?

Comment (LittleFox94) [reply pkg/cloudprovider/provider/anexia/network_provisioning.go]: actually found something to make better with those if statements - see force-pushed changes in a bit :)

File: examples/anexia-machinedeployment.yaml
Comment (LittleFox94) [reply examples/anexia-machinedeployment.yaml]: yea yea xD

File: examples/anexia-machinedeployment.yaml
Comment (LittleFox94) [reply examples/anexia-machinedeployment.yaml]: Changed the comment a bit, yet another version

File: pkg/cloudprovider/provider/anexia/network_provisioning.go
Comment (nachtjasmin) [reply pkg/cloudprovider/provider/anexia/network_provisioning.go:63]: welp, anyway, better than nothing. ^^'

File: pkg/cloudprovider/provider/anexia/network_provisioning.go
Comment (nachtjasmin) [reply pkg/cloudprovider/provider/anexia/network_provisioning.go:82]: And you workaround this possibility with a wonderful explanation inside the comment! It's perfection, thank you! :neodog_heart:

File: pkg/cloudprovider/provider/anexia/types/types.go
Comment (nachtjasmin) [reply pkg/cloudprovider/provider/anexia/types/types.go:82]: > I did not test how this version of machine-controller behaves when reconciling an existing VM though - I would guess nothing happens, but I will check this.

Oh yeah, that'd be interesting. But I think nothing big should happen at all. If you could write a test for that, it'd be greatly appreciated tho. :3

**Review by nachtjasmin — APPROVED** [review]
LGTM would be an understatement tbh. You really did a wonderful job with this PR, nicely done!

(Note: due to the complexity of the components that are involved, I haven't tested this inside a cluster. To be honest, I wouldn't even know where to start. We should write it down somewhere.)

File: pkg/cloudprovider/provider/anexia/types/types.go
Comment (LittleFox94) [reply pkg/cloudprovider/provider/anexia/types/types.go:82]: I did test this manually by changing the machine-controller image in an existing cluster that had an existing `MachineDeployment` and `Machine`

* scaled the MachineDeployment from 1 to 2 replicas -> new Machine successfully created and Node joined the cluster, got warnings for it using the old vlanID config as expected
* scaled the MachineDeployment from 2 to 1 replicas -> the old existing Machine was chosen for deletion and that worked as expected

I don't see any benefit of putting this into an automated test, since that's only a scenario while migrating and this manual test + our knowledge about the behavior tell us "there's nothing to worry here"

**Comment by LittleFox94** [comment]
> It's missing some things we probably want (I'll keep this list updated as I remember things):
> 
> * validate configured prefixes to all be in the configured VLAN

This validation was easy enough to add - but breaks the tests. Fixing the tests would require quite a large refactoring, which I just don't wanna do rn .. the Engine should already complain about bad configurations, just less nice. Imho ok as-is, comments about this are appreciated though.

File: pkg/cloudprovider/provider/anexia/types/types.go
Comment (nachtjasmin) [reply pkg/cloudprovider/provider/anexia/types/types.go:82]: thanks mommy! I appwove this code now, you did such a great job! :pleading_face:
thank you for bapping the machine derployments awawa

Comment (LittleFox94) [reply pkg/cloudprovider/provider/anexia/types/types.go:82]: ... ok I had that coming xD

**Comment by LittleFox94** [comment]
closing this for kubermatic#1845

**Pull request closed without merging**
